File: subset sum equal to target/Tabulation.py
if __name__ == '__main__':
    arr = [1,2,3,4]
    target_sum = 4

    size = len(arr)
    dp = [ [False for _ in range(target_sum+1)] for _ in range(size)]

    for i in range(size):
        dp[i][0] = True 
    
    dp[0][arr[0]] = True;

    for index in range(1,size):
        for target in range(1,target_sum+1):
            nonTake = dp[index-1][target]
            take = False
            if target >= arr[index] :take = dp[index - 1][target - arr[index]]

            dp[index][target] = nonTake or take

    
    print(dp[size-1][target_sum])

File: subset sum equal to target/Tabulation.java

class Tabulation {

    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4};
        int target = 4;

        int size = arr.length;

        boolean[][] dp = new boolean[size][target + 1];

        for (int index = 0; index < size; index++) {
            dp[index][0] = true;
        }

        dp[0][arr[0]] = true;

        for (int i = 1; i < size; i++) {
            for (int j = 1; j < target + 1; j++) {
                boolean nonTake = dp[i - 1][j];
                boolean take = false;
                if (j >= arr[i]) {
                    take =  dp[i-1][j - arr[i] ] ;
                }

                dp[i][j] = (nonTake || take);
            }
        }


        System.out.println(dp[size-1][target]);
    }
}
